File: leetcode/2130.cpp
#include <iostream>
#include <cstring>
#include <stdlib.h>
#include <vector>


using std::vector;
using std::cout;
using std::cin;
using std::endl;
using std::string;
using std::max;


// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    int pairSum(ListNode* head) {
        ListNode* thead = head;
        int n=0;
        while(thead){
            n++;
            thead = thead->next;
        }
        int half = (int) n/2;
        int stack[half];
        int ans = 0;
        int i = 0,j = half-1;
        thead = head;
        while(thead){
            if(i<half){
                stack[i++] = thead->val;
            }else{
                ans = max(ans,stack[j--]+thead->val);
            }
            thead = thead->next;
        }
        return ans;
    }
};

File: leetcode/2130.py
# https://leetcode.com/problems/maximum-twin-sum-of-a-linked-list/description/
from typing import *
from USEFUL_CODES.LinkedList import *


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        thead = head
        n = 0
        while thead:
            thead = thead.next
            n += 1

        nodes = []
        ans = 0
        thead = head
        i = 0
        while thead:
            if i < n // 2:
                nodes.append(thead.val)
            else:
                ans = max(ans, nodes.pop() + thead.val)
            thead = thead.next
            i += 1

        return ans
